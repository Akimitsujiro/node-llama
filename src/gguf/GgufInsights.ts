import {getGgufMetadataLlmData} from "./ggufParser/utils/getGgufMetadataLlmData.js";
import {GgufMetadata} from "./ggufParser/types/GgufMetadataTypes.js";

export class GgufInsights {
    public readonly metadata: GgufMetadata;
    public readonly metadataSize: number;

    public constructor({
        metadata,
        metadataSize
    }: {
        metadata: GgufMetadata,
        metadataSize: number
    }) {
        this.metadata = metadata;
        this.metadataSize = metadataSize;
    }

    /**
     * fp16 k,v matrices
     */
    public get kvMatrices() {
        // 2 bytes each * 2 key and value
        const llmData = getGgufMetadataLlmData(this.metadata);
        return (
            2 * 2 *
            (llmData.context_length ?? 1) *
            (llmData.block_count ?? 1) *
            (llmData.embedding_length ?? 1) *
            (llmData.attention?.head_count_kv ?? 1) /
            (llmData.attention?.head_count ?? 1)
        );
    }

    /**
     * This amount is the overhead + tensors in memory
     */
    public get graphSize() {
        // TODO: get this from the llama.cpp's graph calculations instead of
        // estimating it's 1/6 * kv_cache_size * num_gqa
        const llmData = getGgufMetadataLlmData(this.metadata);
        return (
            (llmData.attention?.head_count_kv ?? 1) /
            (llmData.attention?.head_count ?? 1)
        ) * this.kvMatrices / 6;
    }

    public get VRAMUsage() {
        return this.graphSize + this.kvMatrices + this.metadataSize;
    }
}
