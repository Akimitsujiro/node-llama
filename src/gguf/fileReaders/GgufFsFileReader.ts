import fs from "node:fs/promises";
import retry from "async-retry";
import {withLock} from "lifecycle-utils";
import {GgufReadOffset} from "../utils/GgufReadOffset.js";
import {defaultExtraAllocationSize, ggufDefaultRetryOptions} from "../consts.js";
import {GgufFileReader} from "./GgufFileReader.js";

type GgufFsFileReaderOptions = {
    filePath: string,
    retryOptions?: retry.Options
};

export class GgufFsFileReader extends GgufFileReader {
    public readonly filePath: string;
    public readonly retryOptions: retry.Options;

    public constructor({filePath, retryOptions = ggufDefaultRetryOptions}: GgufFsFileReaderOptions) {
        super();
        this.filePath = filePath;
        this.retryOptions = retryOptions;
    }

    public async readByteRange(offset: number | GgufReadOffset, length: number) {
        const readOffset = GgufReadOffset.resolveReadOffset(offset);
        const endOffset = readOffset.offset + length;

        if (endOffset >= this._buffer.length)
            await this._readToExpandBufferUpToOffset(endOffset);

        const res = this._buffer.subarray(readOffset.offset, endOffset);
        readOffset.moveBy(length);
        return res;
    }

    private async _readToExpandBufferUpToOffset(endOffset: number, extraAllocationSize: number = defaultExtraAllocationSize) {
        return await withLock(this, "modifyBuffer", async () => {
            if (endOffset < this._buffer.length)
                return;

            const missingBytesBuffer = await retry(async () => {
                return await this._readByteRange(this._buffer.length, endOffset + extraAllocationSize - this._buffer.length);
            }, this.retryOptions);

            this._addToBuffer(missingBytesBuffer);
        });
    }

    private async _readByteRange(start: number, length: number) {
        const fd = await fs.open(this.filePath, "r");
        try {
            const buffer = Buffer.alloc(length);
            await fd.read(buffer, 0, length, start);
            return buffer;
        } finally {
            await fd.close();
        }
    }
}
