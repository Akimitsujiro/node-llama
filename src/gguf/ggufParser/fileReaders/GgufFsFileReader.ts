import fs from "node:fs/promises";
import retry from "async-retry";
import {withLock} from "lifecycle-utils";
import {GgufReadOffset} from "../utils/GgufReadOffset.js";
import {GgufBaseFileReader, ALLOCATION_SIZE} from "./GgufBaseFileReader.js";

type GgufFsFileReaderOptions = {
    filePath: string,
    retryOptions?: retry.Options
};

const defaultRetryOptions: retry.Options = {
    retries: 10,
    factor: 2,
    minTimeout: 1000,
    maxTimeout: 1000 * 16
} as const;

export class GgufFsFileReader extends GgufBaseFileReader {
    public readonly filePath: string;
    public readonly retryOptions: retry.Options;

    public constructor({filePath, retryOptions = defaultRetryOptions}: GgufFsFileReaderOptions) {
        super();
        this.filePath = filePath;
        this.retryOptions = retryOptions;
    }

    public async readByteRange(offset: number | GgufReadOffset, length: number) {
        const readOffset = GgufReadOffset.resolveReadOffset(offset);
        const endOffset = readOffset.offset + length;

        if (endOffset >= this._buffer.length)
            await this._readToExpandBufferUpToOffset(endOffset);

        const res = this._buffer.subarray(readOffset.offset, endOffset);
        readOffset.moveBy(length);
        return res;
    }

    private async _readToExpandBufferUpToOffset(endOffset: number, extraAllocationSize: number = ALLOCATION_SIZE) {
        return await withLock(this, "modifyBuffer", async () => {
            if (endOffset < this._buffer.length)
                return;

            const missingBytesBuffer = await retry(async () => {
                return await this._readByteRange(this._buffer.length, endOffset + extraAllocationSize - this._buffer.length);
            }, this.retryOptions);

            this._addToBuffer(missingBytesBuffer);
        });
    }

    private async _readByteRange(start: number, length: number) {
        const fd = await fs.open(this.filePath, "r");
        try {
            const buffer = Buffer.alloc(length);
            await fd.read(buffer, 0, length, start);
            return buffer;
        } finally {
            await fd.close();
        }
    }
}
