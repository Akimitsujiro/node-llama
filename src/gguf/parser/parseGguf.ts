import {InvalidGgufMagicError} from "../errors/InvalidGgufMagicError.js";
import {getConsoleLogPrefix} from "../../utils/getConsoleLogPrefix.js";
import {UnsupportedError} from "../../utils/UnsupportedError.js";
import {GgufReadOffset} from "../utils/GgufReadOffset.js";
import {GgufFileReader, valueTypeToBytesToRead} from "../fileReaders/GgufFileReader.js";
import {GgufFileInfo, GgufVersionParserOptions, GgufVersionParserResult} from "../types/GgufFileInfoTypes.js";
import {GgufV2Parser} from "./GgufV2Parser.js";

const ggufMagic = "GGUF";

export async function parseGguf({
    fileReader,
    readTensorInfo = true,
    ignoreKeys = [],
    logWarnings = true
}: {
    fileReader: GgufFileReader,
    readTensorInfo?: boolean,
    ignoreKeys?: string[],
    logWarnings?: boolean
}): Promise<GgufFileInfo> {
    const readOffset = new GgufReadOffset(0);
    const magicAndVersion = await parseMagicAndVersion(fileReader, readOffset);
    const gguifInfo = await parseGgufUsingASpecificVersionParser({
        fileReader,
        readTensorInfo,
        ignoreKeys,

        version: magicAndVersion.version,
        readOffset,
        logWarnings
    });

    return {
        version: magicAndVersion.version,
        tensorCount: gguifInfo.tensorCount,
        metadata: gguifInfo.metadata,
        tensorInfo: gguifInfo.tensorInfo,
        metadataSize: gguifInfo.metadataSize,
        tensorInfoSize: gguifInfo.tensorInfoSize
    };
}

async function parseMagicAndVersion(fileReader: GgufFileReader, readOffset: GgufReadOffset) {
    const fileMagicBytes = await fileReader.readByteRange(readOffset, valueTypeToBytesToRead.uint8 * ggufMagic.length);
    const fileMagicText = String.fromCharCode(...fileMagicBytes);

    if (fileMagicText !== ggufMagic)
        throw new InvalidGgufMagicError(ggufMagic, fileMagicText);

    const version = await fileReader.readUint32(readOffset);

    return {
        magic: ggufMagic,
        version
    };
}

async function parseGgufUsingASpecificVersionParser(
    specificVersionParserOptions: GgufVersionParserOptions
): Promise<GgufVersionParserResult> {
    switch (specificVersionParserOptions.version) {
        case 1:
            throw new UnsupportedError("GGUF version 1 is not supported by llama.cpp anymore");

        case 2:
        case 3:
            return await (new GgufV2Parser(specificVersionParserOptions)).parse();

        default:
            if (specificVersionParserOptions.logWarnings)
                console.warn(
                    getConsoleLogPrefix() +
                    `Unsupported GGUF version: ${specificVersionParserOptions.version}. Parsing it as version 3`
                );

            return await (new GgufV2Parser(specificVersionParserOptions)).parse();
    }
}
