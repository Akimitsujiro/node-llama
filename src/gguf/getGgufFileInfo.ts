import retry from "async-retry";
import {GgufParser} from "./ggufParser/GgufParser.js";
import {GgufNetworkFetchFileReader} from "./ggufParser/fileReaders/GgufNetworkFetchFileReader.js";
import {GgufFsFileReader} from "./ggufParser/fileReaders/GgufFsFileReader.js";
import {ggufDefaultRetryOptions} from "./consts.js";


/**
 * Parse a GGUF file and return its metadata and tensor info (unless `readTensorInfo` is set to `false`)
 */
export async function getGgufFileInfo(pathOrUrl: string, {
    readTensorInfo = true,
    sourceType,
    retryOptions = ggufDefaultRetryOptions,
    ignoreKeys = []
}: {
    readTensorInfo?: boolean,
    sourceType?: "network" | "filesystem",
    retryOptions?: retry.Options,
    ignoreKeys?: string[]
} = {}) {
    function createFileReader() {
        if (sourceType === "network" || (sourceType == null && (pathOrUrl.startsWith("http://") || pathOrUrl.startsWith("https://")))) {
            return new GgufNetworkFetchFileReader({
                url: pathOrUrl,
                retryOptions: retryOptions
            });
        } else if (sourceType === "filesystem" || sourceType == null) {
            return new GgufFsFileReader({
                filePath: pathOrUrl,
                retryOptions: retryOptions
            });
        }

        void (sourceType satisfies never);
        throw new Error(`Unsupported sourceType: ${sourceType}`);
    }

    const fileReader = createFileReader();
    const parser = new GgufParser({
        fileReader,
        ignoreKeys,
        readTensorInfo
    });

    return await parser.parseFileInfo();
}
