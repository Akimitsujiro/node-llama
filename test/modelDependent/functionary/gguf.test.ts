import {describe, expect, it, test} from "vitest";
import {GgufFsFileReader} from "../../../src/gguf/fileReaders/GgufFsFileReader.js";
import {parseGguf} from "../../../src/gguf/parser/parseGguf.js";
import {getModelFile} from "../../utils/modelFiles.js";
import {GgufInsights} from "../../../src/gguf/GgufInsights.js";
import {getTestLlama} from "../../utils/getTestLlama.js";
import {getGgufFileInfo} from "../../../src/gguf/getGgufFileInfo.js";
import {simplifyGgufInfoForTestSnapshot} from "../../utils/helpers/simplifyGgufInfoForTestSnapshot.js";

describe("GGUF Parser", async () => {
    const modelPath = await getModelFile("functionary-small-v2.2.q4_0.gguf");

    test("Magic should be GGUF local model", async () => {
        const fileReader = new GgufFsFileReader({filePath: modelPath});
        const magic = await fileReader.readByteRange(0, 4);
        const magicText = String.fromCharCode(...magic);

        expect(magicText).toBe("GGUF");
    });

    it("should parse local gguf model", async () => {
        const fileReader = new GgufFsFileReader({filePath: modelPath});

        const metadata = await parseGguf({
            fileReader: fileReader
        });
        expect(metadata.tensorInfo!.length).to.be.eql(Number(metadata.tensorCount));

        expect(simplifyGgufInfoForTestSnapshot(metadata)).toMatchSnapshot();
    });

    it("should calculate GGUF VRAM Usage", async () => {
        const fileReader = new GgufFsFileReader({filePath: modelPath});

        const metadata = await parseGguf({
            fileReader: fileReader
        });

        const ggufInsights = new GgufInsights(metadata);

        const llama = await getTestLlama();
        const model = await llama.loadModel({
            modelPath: modelPath
        });

        const usedRam = llama.getVramState().used;

        expect(ggufInsights.VRAMUsage).toMatchInlineSnapshot("4474643028.666667");
        expect(usedRam).to.be.gte(3.5 * Math.pow(1024, 3));
        expect(usedRam).to.be.lte(4.5 * Math.pow(1024, 3));
        await model.dispose();
    });

    it("should fetch GGUF metadata", async () => {
        const ggufMetadataParseResult = await getGgufFileInfo(modelPath);

        expect(simplifyGgufInfoForTestSnapshot(ggufMetadataParseResult)).toMatchSnapshot();

        const insights = new GgufInsights(ggufMetadataParseResult);
        expect(insights.VRAMUsage).toMatchInlineSnapshot("4474643028.666667");
    });
});
