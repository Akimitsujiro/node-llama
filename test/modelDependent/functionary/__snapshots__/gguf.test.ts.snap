// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GGUF Parser > should fetch GGUF metadata 1`] = `
{
  "metadata": {
    "general": {
      "architecture": "llama",
      "file_type": 2,
      "name": "workspace",
      "quantization_version": 2,
    },
    "llama": {
      "attention": {
        "head_count": 32,
        "head_count_kv": 8,
        "layer_norm_rms_epsilon": 0.000009999999747378752,
      },
      "block_count": 32,
      "context_length": 32768,
      "embedding_length": 4096,
      "feed_forward_length": 14336,
      "rope": {
        "dimension_count": 128,
        "freq_base": 10000,
      },
    },
    "tokenizer": {
      "chat_template": "{% for message in messages %}
{% if message['role'] == 'user' or message['role'] == 'system' %}
{{ '<|from|>' + message['role'] + '
<|recipient|>all
<|content|>' + message['content'] + '
' }}{% elif message['role'] == 'tool' %}
{{ '<|from|>' + message['name'] + '
<|recipient|>all
<|content|>' + message['content'] + '
' }}{% else %}
{% set contain_content='no'%}
{% if message['content'] is not none %}
{{ '<|from|>assistant
<|recipient|>all
<|content|>' + message['content'] }}{% set contain_content='yes'%}
{% endif %}
{% if 'tool_calls' in message and message['tool_calls'] is not none %}
{% for tool_call in message['tool_calls'] %}
{% set prompt='<|from|>assistant
<|recipient|>' + tool_call['function']['name'] + '
<|content|>' + tool_call['function']['arguments'] %}
{% if loop.index == 1 and contain_content == "no" %}
{{ prompt }}{% else %}
{{ '
' + prompt}}{% endif %}
{% endfor %}
{% endif %}
{{ '<|stop|>
' }}{% endif %}
{% endfor %}
{% if add_generation_prompt %}{{ '<|from|>assistant
<|recipient|>' }}{% endif %}",
      "ggml": {
        "bos_token_id": 1,
        "eos_token_id": 2,
        "model": "llama",
        "padding_token_id": 2,
        "scores": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "token_type": [
          2,
          3,
          3,
          6,
          6,
          6,
          6,
          6,
          6,
          6,
        ],
        "tokens": [
          "<unk>",
          "<s>",
          "</s>",
          "<0x00>",
          "<0x01>",
          "<0x02>",
          "<0x03>",
          "<0x04>",
          "<0x05>",
          "<0x06>",
        ],
        "unknown_token_id": 0,
      },
    },
  },
  "metadataSize": 718762,
  "tensorCount": 291,
  "tensorInfo": [
    {
      "dimensions": [
        4096,
        32004,
      ],
      "ggmlType": 2,
      "name": "token_embd.weight",
      "offset": 0,
    },
    {
      "dimensions": [
        4096,
      ],
      "ggmlType": 0,
      "name": "blk.0.attn_norm.weight",
      "offset": 73737216,
    },
    {
      "dimensions": [
        14336,
        4096,
      ],
      "ggmlType": 2,
      "name": "blk.0.ffn_down.weight",
      "offset": 73753600,
    },
    {
      "dimensions": [
        4096,
        14336,
      ],
      "ggmlType": 2,
      "name": "blk.0.ffn_gate.weight",
      "offset": 106783744,
    },
  ],
  "tensorInfoSize": 17286,
  "version": 3,
}
`;

exports[`GGUF Parser > should parse local gguf model 1`] = `
{
  "metadata": {
    "general": {
      "architecture": "llama",
      "file_type": 2,
      "name": "workspace",
      "quantization_version": 2,
    },
    "llama": {
      "attention": {
        "head_count": 32,
        "head_count_kv": 8,
        "layer_norm_rms_epsilon": 0.000009999999747378752,
      },
      "block_count": 32,
      "context_length": 32768,
      "embedding_length": 4096,
      "feed_forward_length": 14336,
      "rope": {
        "dimension_count": 128,
        "freq_base": 10000,
      },
    },
    "tokenizer": {
      "chat_template": "{% for message in messages %}
{% if message['role'] == 'user' or message['role'] == 'system' %}
{{ '<|from|>' + message['role'] + '
<|recipient|>all
<|content|>' + message['content'] + '
' }}{% elif message['role'] == 'tool' %}
{{ '<|from|>' + message['name'] + '
<|recipient|>all
<|content|>' + message['content'] + '
' }}{% else %}
{% set contain_content='no'%}
{% if message['content'] is not none %}
{{ '<|from|>assistant
<|recipient|>all
<|content|>' + message['content'] }}{% set contain_content='yes'%}
{% endif %}
{% if 'tool_calls' in message and message['tool_calls'] is not none %}
{% for tool_call in message['tool_calls'] %}
{% set prompt='<|from|>assistant
<|recipient|>' + tool_call['function']['name'] + '
<|content|>' + tool_call['function']['arguments'] %}
{% if loop.index == 1 and contain_content == "no" %}
{{ prompt }}{% else %}
{{ '
' + prompt}}{% endif %}
{% endfor %}
{% endif %}
{{ '<|stop|>
' }}{% endif %}
{% endfor %}
{% if add_generation_prompt %}{{ '<|from|>assistant
<|recipient|>' }}{% endif %}",
      "ggml": {
        "bos_token_id": 1,
        "eos_token_id": 2,
        "model": "llama",
        "padding_token_id": 2,
        "scores": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "token_type": [
          2,
          3,
          3,
          6,
          6,
          6,
          6,
          6,
          6,
          6,
        ],
        "tokens": [
          "<unk>",
          "<s>",
          "</s>",
          "<0x00>",
          "<0x01>",
          "<0x02>",
          "<0x03>",
          "<0x04>",
          "<0x05>",
          "<0x06>",
        ],
        "unknown_token_id": 0,
      },
    },
  },
  "metadataSize": 718762,
  "tensorCount": 291,
  "tensorInfo": [
    {
      "dimensions": [
        4096,
        32004,
      ],
      "ggmlType": 2,
      "name": "token_embd.weight",
      "offset": 0,
    },
    {
      "dimensions": [
        4096,
      ],
      "ggmlType": 0,
      "name": "blk.0.attn_norm.weight",
      "offset": 73737216,
    },
    {
      "dimensions": [
        14336,
        4096,
      ],
      "ggmlType": 2,
      "name": "blk.0.ffn_down.weight",
      "offset": 73753600,
    },
    {
      "dimensions": [
        4096,
        14336,
      ],
      "ggmlType": 2,
      "name": "blk.0.ffn_gate.weight",
      "offset": 106783744,
    },
  ],
  "tensorInfoSize": 17286,
  "version": 3,
}
`;
